up-streamer/src/ustreamer.rs

```diff
@@ -0,0 +1,974 @@
+            registered_forwarding_rules: Mutex::new(HashMap::new()),
+            transport_forwarders: Mutex::new(HashMap::new()),
+            transport_forwarder_count: Mutex::new(HashMap::new()),
+            in_transport_out_authorities: Mutex::new(HashMap::new()),
+            message_queue_size,
+        }
+    }
+
+    fn uauthority_to_uuri(authority: UAuthority) -> UUri {
+        UUri {
+            authority: Some(authority).into(),
+            ..Default::default()
+        }
+    }
+
+    #[inline(always)]
+    fn forwarding_id(r#in: &Endpoint, out: &Endpoint) -> String {
+        format!(
+            "[in.name: {}, in.authority: {:?} ; out.name: {}, out.authority: {:?}]",
+            r#in.name, r#in.authority, out.name, out.authority
+        )
+    }
+
+    #[inline(always)]
+    fn fail_due_to_same_authority(&self, r#in: &Endpoint, out: &Endpoint) -> Result<(), UStatus> {
+        let err = Err(UStatus::fail_with_code(
+            UCode::INVALID_ARGUMENT,
+            format!(
+                "{} are the same. Unable to delete.",
+                Self::forwarding_id(r#in, out)
+            ),
+        ));
+        error!(
+            "{}:{}:{} Deleting forwarding rule failed: {:?}",
+            self.name, USTREAMER_TAG, USTREAMER_FN_ADD_FORWARDING_RULE_TAG, err
+        );
+        err
+    }
+
+    /// Adds a forwarding rule to the [`UStreamer`] based on an in [`Endpoint`][crate::Endpoint] and an
+    /// out [`Endpoint`][crate::Endpoint]
+    ///
+    /// Works for any [`UMessage`][up_rust::UMessage] type which has a destination / sink contained
+    /// in its attributes, i.e.
+    /// * [`UMessageType::UMESSAGE_TYPE_NOTIFICATION`][up_rust::UMessageType::UMESSAGE_TYPE_NOTIFICATION]
+    /// * [`UMessageType::UMESSAGE_TYPE_REQUEST`][up_rust::UMessageType::UMESSAGE_TYPE_REQUEST]
+    /// * [`UMessageType::UMESSAGE_TYPE_RESPONSE`][up_rust::UMessageType::UMESSAGE_TYPE_RESPONSE]
+    ///
+    /// # Parameters
+    ///
+    /// * `in` - [`Endpoint`][crate::Endpoint] we will bridge _from_
+    /// * `out` - [`Endpoint`][crate::Endpoint] we will bridge _onto_
+    ///
+    /// # Errors
+    ///
+    /// If unable to add this forwarding rule, we return a [`UStatus`][up_rust::UStatus] noting
+    /// the error.
+    ///
+    /// Typical errors include
+    /// * already have this forwarding rule registered
+    /// * attempting to forward onto the same [`Endpoint`][crate::Endpoint]
+    pub async fn add_forwarding_rule(&self, r#in: Endpoint, out: Endpoint) -> Result<(), UStatus> {
+        debug!(
+            "{}:{}:{} Adding forwarding rule for {}",
+            self.name,
+            USTREAMER_TAG,
+            USTREAMER_FN_ADD_FORWARDING_RULE_TAG,
+            Self::forwarding_id(&r#in, &out)
+        );
+
+        if r#in.authority == out.authority {
+            return self.fail_due_to_same_authority(&r#in, &out);
+        }
+
+        let in_comparable_transport = ComparableTransport::new(r#in.transport.clone());
```

Comment:
looks like we could make use of something like a parameterized `ComparableTraitObjectReference<T>` in up-rust?
:thinking:

Reply:
Forget about it, it seems that this cannot be done because of the characteristics of trait objects vs types